flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/TableAggregateITCase.scala

```diff
@@ -0,0 +1,227 @@
+      "5,14,1",
+      "5,13,2",
+      "6,21,0",
+      "6,20,1",
+      "6,19,2"
+    ).sorted
+    assertEquals(expected, retracted)
+  }
+
+  @Test
+  def testTop3InAccRetractModeWithoutGroupBy(): Unit = {
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
+    env.setStateBackend(getStateBackend)
+    val tEnv = TableEnvironment.getTableEnvironment(env)
+    StreamITCase.clear
+
+    val top3 = new TopN(3)
+    val source = StreamTestData.get3TupleDataStream(env).toTable(tEnv, 'a, 'b, 'c)
+    val t = source
+      .flatAggregate(top3("1".cast(Types.INT), 'a.cast(Types.LONG)))
+      .select('_1, '_2, '_3)
+      .as('category, 'v, 'rank)
+
+    val results = t.toRetractStream[Row](queryConfig)
+    results.addSink(new StreamITCase.RetractingSink).setParallelism(1)
+    env.execute()
+
+    val expected = mutable.MutableList(
+      "1,21,0",
+      "1,20,1",
+      "1,19,2")
+    assertEquals(expected.sorted, StreamITCase.retractedResults.sorted)
+  }
+
+  @Test
+  def testTop3InAccRetractModeWithGroupBy(): Unit = {
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
+    env.setStateBackend(getStateBackend)
+    val tEnv = TableEnvironment.getTableEnvironment(env)
+    StreamITCase.clear
+
+    val top3 = new TopN(3)
+    val source = StreamTestData.get3TupleDataStream(env).toTable(tEnv, 'a, 'b, 'c)
+    val t = source.groupBy('b)
+      .flatAggregate(top3('b.cast(Types.INT), 'a.cast(Types.LONG)))
+      .select('_1, '_2, '_3)
+
+    val results = t.toRetractStream[Row](queryConfig)
+    results.addSink(new StreamITCase.RetractingSink)
+    env.execute()
+
+    val expected = mutable.MutableList(
+      "1,1,0",
+      "2,3,0",
+      "2,2,1",
+      "3,6,0",
+      "3,5,1",
+      "3,4,2",
+      "4,10,0",
+      "4,9,1",
+      "4,8,2",
+      "5,15,0",
+      "5,14,1",
+      "5,13,2",
+      "6,21,0",
+      "6,20,1",
+      "6,19,2")
+    assertEquals(expected.sorted, StreamITCase.retractedResults.sorted)
+  }
+
+  // test with retraction input
+  @Test
+  def testTop3InAccRetractModeWithRetractInput(): Unit = {
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
+    env.setStateBackend(getStateBackend)
+    val tEnv = TableEnvironment.getTableEnvironment(env)
+    StreamITCase.clear
+
+    val top3 = new Top3WithRetractInput
+    val source = StreamTestData.get3TupleDataStream(env).toTable(tEnv, 'a, 'b, 'c)
+    val t = source
+      .groupBy('b)
+      .select("1" as 'a, 'b, 'a.count as 'count)
+      .groupBy('a)
+      .flatAggregate(top3('b.cast(Types.INT), 'count))
+      .select('_1, '_2, '_3)
+
+    val results = t.toRetractStream[Row](queryConfig)
+    results.addSink(new StreamITCase.RetractingSink)
+    env.execute()
+
+    val expected = mutable.MutableList(
+      "1,6,0",
+      "1,5,1",
+      "1,4,2")
+    assertEquals(expected.sorted, StreamITCase.retractedResults.sorted)
+  }
+
+  @Test(expected = classOf[ValidationException])
+  def testEmitMethod(): Unit = {
```

Comment:
Move to TableAggregateValidationTest

Reply:
It seems we can't move the test into `TableAggregateValidationTest` as the exception is thrown during translate to plan.